lib/puppet/pops/types/p_sem_ver_type.rb

```diff
@@ -0,0 +1,121 @@
+module Puppet::Pops
+module Types
+
+# A Puppet Language Type that exposes the {{Semantic::Version}} and {{Semantic::VersionRange}}.
+# The version type is parameterized with version ranges.
+#
+# @api public
+class PSemVerType < PScalarType
+  attr_reader :ranges
+
+  def initialize(*ranges)
+    ranges = ranges.map { |range| range.is_a?(Semantic::VersionRange) ? range : Semantic::VersionRange.parse(range) }
+    ranges = merge_ranges(ranges) if ranges.size > 1
+    @ranges = ranges
+  end
+
+  def instance?(o, guard = nil)
+    o.is_a?(Semantic::Version) && (@ranges.empty? || @ranges.any? {|range| range.include?(o) })
+  end
+
+  def eql?(o)
+    self.class == o.class && @ranges == o.ranges
+  end
+
+  def hash?
+    super ^ @ranges.hash
+  end
+
+  # @api private
+  def self.parts_pattern
+    part = "(?<part>[0-9A-Za-z-]+)"
+    "(?<parts>#{part}(?:\\.\\g<part>)*)"
+  end
+
+  # @api private
+  def self.version_pattern
+    nr = '(?<nr>0|[1-9][0-9]*)'
+    "#{nr}\\.\\g<nr>\\.\\g<nr>(?:-#{parts_pattern})?(?:\\+\\g<parts>)?"
+  end
+
+  # @api private
+  def self.new_function(_, loader)
```

Comment:
The documentation for this needs to be in the `new` function as that is the user facing function atm. We have not other way (yet) to generate documentation for the operations available on types.